### PR TITLE
GoogleFile fix

### LINKED ISSUE #1623: GoogleFile partially broken
When trying to use the GoogleFile put method I get an error due to a missing dependency

```
Ganga In [1]: g = GoogleFile()
Ganga In [2]: g.namePattern='test'
Ganga In [3]: g.localDir = '~/tmp'
Ganga In [4]: g.put()                                                                                    
ERROR    !!Unknown/Unexpected ERROR!!
:
ModuleNotFoundError: No module named 'oauth2client'
```
When just installing this module manually, I get a long list of warnings, but the code actually works.

```
Ganga In [4]: g.put()                                                                                    
WARNING  __init__() takes at most 4 positional arguments (5 given)
INFO     Go to the following link in your browser: https://accounts.google.com/o/oauth2/v2/auth?client_id=776655306197-dirtoquqsm7cpqgepvamofg5t2b5f637.apps.googleusercontent.com&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fdrive.file&access_type=offline&response_type=code
Enter verification code: 4/yAHOqQ5PXNpnpITCTi757SgAlOSiSCp8tjjN-a3bYJZjQRSEymH2OXo
INFO     Your GoogleDrive credentials have been stored in the file /home/egede/gangadir/googlecreddata.pkl and are only readable by you. The file will give permission to modify files in your GoogleDrive. Permission can be revoked by going to "Manage Apps" in your GoogleDrive or by deleting the credentials through the deleteCredentials GoogleFile method.
WARNING  file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth
Traceback (most recent call last):
  File "/home/egede/programming/ganga/lib/python3.7/site-packages/googleapiclient/discovery_cache/__init__.py", line 36, in autodetect
    from google.appengine.api import memcache
ModuleNotFoundError: No module named 'google.appengine'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/egede/programming/ganga/lib/python3.7/site-packages/googleapiclient/discovery_cache/file_cache.py", line 33, in <module>
    from oauth2client.contrib.locked_file import LockedFile
ModuleNotFoundError: No module named 'oauth2client.contrib.locked_file'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/egede/programming/ganga/lib/python3.7/site-packages/googleapiclient/discovery_cache/file_cache.py", line 37, in <module>
    from oauth2client.locked_file import LockedFile
ModuleNotFoundError: No module named 'oauth2client.locked_file'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/egede/programming/ganga/lib/python3.7/site-packages/googleapiclient/discovery_cache/__init__.py", line 41, in autodetect
    from . import file_cache
  File "/home/egede/programming/ganga/lib/python3.7/site-packages/googleapiclient/discovery_cache/file_cache.py", line 41, in <module>
    'file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth')
ImportError: file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth
WARNING  file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth
Traceback (most recent call last):
  File "/home/egede/programming/ganga/lib/python3.7/site-packages/googleapiclient/discovery_cache/__init__.py", line 36, in autodetect
    from google.appengine.api import memcache
ModuleNotFoundError: No module named 'google.appengine'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/egede/programming/ganga/lib/python3.7/site-packages/googleapiclient/discovery_cache/file_cache.py", line 33, in <module>
    from oauth2client.contrib.locked_file import LockedFile
ModuleNotFoundError: No module named 'oauth2client.contrib.locked_file'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/egede/programming/ganga/lib/python3.7/site-packages/googleapiclient/discovery_cache/file_cache.py", line 37, in <module>
    from oauth2client.locked_file import LockedFile
ModuleNotFoundError: No module named 'oauth2client.locked_file'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/egede/programming/ganga/lib/python3.7/site-packages/googleapiclient/discovery_cache/__init__.py", line 41, in autodetect
    from . import file_cache
  File "/home/egede/programming/ganga/lib/python3.7/site-packages/googleapiclient/discovery_cache/file_cache.py", line 41, in <module>
    'file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth')
ImportError: file_cache is unavailable when using oauth2client >= 4.0.0 or google-auth
INFO     File 'test' uploaded succesfully
```
The two problems might be related (I see something about version of oauth2) so they probably need to be fixed together.

### PR DESCRIPTION
**Pull request by DumbMachine**
Aims to fix GoogleFile partially broken #1623
TODO:
- [x] fix auth()
- [x] fix put()
- [x] fix restore()
- [x] fix remove()
- [x] implement get()
- [x] fix internalCopyTo()
- [ ] update tests

### CODE REVIEW

**Comment by egede** [comment]
Remember to update setup.py https://github.com/ganga-devs/ganga/blob/develop/setup.py as well to reflect the dependency.

**Comment by DumbMachine** [comment]
@egede Could you tell me commands where:
- remove() function will be used
- restore() function will be used

**Comment by egede** [comment]
The `remove()` function is implemented in the `IGangaFile` interface 
https://github.com/ganga-devs/ganga/blob/af5e19fc7e6da829ce0b7b37ac1ee67b9ac25c4c/ganga/GangaCore/GPIDev/Adapters/IGangaFile.py#L186 For `restore()` I have no idea.

**Comment by egede** [comment]
Some indirect errors from the CI testing. You should be able to run all these with `pytest` locally as well.

```
==================================== ERRORS ====================================
________ ERROR collecting ganga/GangaCore/test/GPI/TestGangaTutorial.py ________
ganga/GangaCore/test/GPI/TestGangaTutorial.py:2: in <module>
    from GangaTutorial.Lib.PrimeFactorizer import PrimeFactorizer
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaTutorial/Lib/__init__.py:1: in <module>
    from .PrimeFactorizer import *
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaTutorial/Lib/PrimeFactorizer.py:10: in <module>
    from GangaCore.GPIDev.Adapters.IPrepareApp import IPrepareApp
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/GPIDev/Adapters/IPrepareApp.py:11: in <module>
    from GangaCore.GPIDev.Lib.File.File import File, ShareDir
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/GPIDev/Lib/File/__init__.py:11: in <module>
    from GangaCore.GPIDev.Lib.File.GoogleFile import GoogleFile
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:143: in exec_module
    source_stat, co = _rewrite_test(fn, self.config)
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:328: in _rewrite_test
    tree = ast.parse(source, filename=fn)
/usr/lib64/python3.6/ast.py:35: in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
E     File "/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py", line 98
E       creds = flow.run_local_server(port=0)
E                                           ^
E   IndentationError: unindent does not match any outer indentation level
________ ERROR collecting ganga/GangaCore/test/GPI/TestGangaTutorial.py ________
venv/lib/python3.6/site-packages/_pytest/python.py:513: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
venv/lib/python3.6/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/test/GPI/TestGangaTutorial.py:2: in <module>
    from GangaTutorial.Lib.PrimeFactorizer import PrimeFactorizer
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaTutorial/Lib/__init__.py:1: in <module>
    from .PrimeFactorizer import *
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaTutorial/Lib/PrimeFactorizer.py:14: in <module>
    from GangaCore.GPIDev.Lib.File import File, ShareDir
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/GPIDev/Lib/File/__init__.py:11: in <module>
    from GangaCore.GPIDev.Lib.File.GoogleFile import GoogleFile
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:143: in exec_module
    source_stat, co = _rewrite_test(fn, self.config)
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:328: in _rewrite_test
    tree = ast.parse(source, filename=fn)
/usr/lib64/python3.6/ast.py:35: in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
E     File "/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py", line 98
E       creds = flow.run_local_server(port=0)
E                                           ^
E   IndentationError: unindent does not match any outer indentation level
___________ ERROR collecting ganga/GangaCore/test/Unit/TestProxy.py ____________
ganga/GangaCore/test/Unit/TestProxy.py:4: in <module>
    from GangaCore.GPIDev.Lib.File import LocalFile
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/GPIDev/Lib/File/__init__.py:11: in <module>
    from GangaCore.GPIDev.Lib.File.GoogleFile import GoogleFile
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:143: in exec_module
    source_stat, co = _rewrite_test(fn, self.config)
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:328: in _rewrite_test
    tree = ast.parse(source, filename=fn)
/usr/lib64/python3.6/ast.py:35: in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
E     File "/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py", line 98
E       creds = flow.run_local_server(port=0)
E                                           ^
E   IndentationError: unindent does not match any outer indentation level
___________ ERROR collecting ganga/GangaCore/test/Unit/TestProxy.py ____________
venv/lib/python3.6/site-packages/_pytest/python.py:513: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
venv/lib/python3.6/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/test/Unit/TestProxy.py:4: in <module>
    from GangaCore.GPIDev.Lib.File import LocalFile
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/GPIDev/Lib/File/__init__.py:11: in <module>
    from GangaCore.GPIDev.Lib.File.GoogleFile import GoogleFile
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:143: in exec_module
    source_stat, co = _rewrite_test(fn, self.config)
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:328: in _rewrite_test
    tree = ast.parse(source, filename=fn)
/usr/lib64/python3.6/ast.py:35: in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
E     File "/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py", line 98
E       creds = flow.run_local_server(port=0)
E                                           ^
E   IndentationError: unindent does not match any outer indentation level
_____ ERROR collecting ganga/GangaCore/test/Unit/Notebook/TestNotebook.py ______
ganga/GangaCore/test/Unit/Notebook/TestNotebook.py:13: in <module>
    from GangaCore.Lib.Notebook import Notebook
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/Lib/Notebook/__init__.py:2: in <module>
    from .Notebook import Notebook
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/Lib/Notebook/Notebook.py:14: in <module>
    from GangaCore.GPIDev.Lib.File import FileUtils
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/GPIDev/Lib/File/__init__.py:11: in <module>
    from GangaCore.GPIDev.Lib.File.GoogleFile import GoogleFile
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:143: in exec_module
    source_stat, co = _rewrite_test(fn, self.config)
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:328: in _rewrite_test
    tree = ast.parse(source, filename=fn)
/usr/lib64/python3.6/ast.py:35: in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
E     File "/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py", line 98
E       creds = flow.run_local_server(port=0)
E                                           ^
E   IndentationError: unindent does not match any outer indentation level
_____ ERROR collecting ganga/GangaCore/test/Unit/Notebook/TestNotebook.py ______
venv/lib/python3.6/site-packages/_pytest/python.py:513: in _importtestmodule
    mod = self.fspath.pyimport(ensuresyspath=importmode)
venv/lib/python3.6/site-packages/py/_path/local.py:701: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/test/Unit/Notebook/TestNotebook.py:13: in <module>
    from GangaCore.Lib.Notebook import Notebook
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/Lib/Notebook/__init__.py:2: in <module>
    from .Notebook import Notebook
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/Lib/Notebook/Notebook.py:14: in <module>
    from GangaCore.GPIDev.Lib.File import FileUtils
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:152: in exec_module
    exec(co, module.__dict__)
ganga/GangaCore/GPIDev/Lib/File/__init__.py:11: in <module>
    from GangaCore.GPIDev.Lib.File.GoogleFile import GoogleFile
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:143: in exec_module
    source_stat, co = _rewrite_test(fn, self.config)
venv/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:328: in _rewrite_test
    tree = ast.parse(source, filename=fn)
/usr/lib64/python3.6/ast.py:35: in parse
    return compile(source, filename, mode, PyCF_ONLY_AST)
E     File "/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py", line 98
E       creds = flow.run_local_server(port=0)
E                                           ^
E   IndentationError: unindent does not match any outer indentation level
=============================== warnings summary ===============================
```

**Comment by DumbMachine** [comment]
@egede can we remove this from the schema?
https://github.com/ganga-devs/ganga/blob/d057b9e9f7fce0713190b4597c7e7ef8580608e3/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py#L55
This is no longer used to download the files. I have used the recommended way of downloading files from the drive. Since this is no longer required, I assume its removal is not a problem.

This also requires a change,
https://github.com/ganga-devs/ganga/blob/d057b9e9f7fce0713190b4597c7e7ef8580608e3/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py#L59
the files are now referred by `name`. This means the `title` should become `name`.

**Comment by egede** [comment]
> @egede can we remove this from the schema?
> 
> https://github.com/ganga-devs/ganga/blob/d057b9e9f7fce0713190b4597c7e7ef8580608e3/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py#L55
> 
> This is no longer used to download the files. I have used the recommended way of downloading files from the drive. Since this is no longer required, I assume its removal is not a problem.

Users might want to know how to access the file outside of Ganga. for this the `IGangaFile` interface expects that the object has an `accessURL` method, https://github.com/ganga-devs/ganga/blob/af5e19fc7e6da829ce0b7b37ac1ee67b9ac25c4c/ganga/GangaCore/GPIDev/Adapters/IGangaFile.py#L192 . If you can create that link dynamically from the information that is otherwise persisted in the GoogleFile object, then you can delete this from the schema. 

> 
> This also requires a change,
> 
> https://github.com/ganga-devs/ganga/blob/d057b9e9f7fce0713190b4597c7e7ef8580608e3/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py#L59
> 
> the files are now referred by `name`. This means the `title` should become `name`.

Yes, feel free to change this one.

**Comment by egede** [comment]
Just let us know when you would like a review.

**Comment by DumbMachine** [comment]
Uploading/deleting/restoring to drive is functional now. I'm currently doing some refactoring

**Comment by DumbMachine** [comment]
I think this is ready for review and testing now. 
Are going to continue with the [old test](https://github.com/ganga-devs/ganga/blob/80bc4d9a6b5cb45bd72647236a782808bb4364e2/ganga/GangaCore/old_test/GPI/TestGoogleFile.py) or write a new one?

**Review by egede — COMMENTED** [review]
Should there not be some change in the dependencies of Ganga as well?

**Comment by DumbMachine** [comment]
> Should there not be some change in the dependencies of Ganga as well?

These libraries will have to be added to https://github.com/ganga-devs/ganga/blob/develop/setup.py:
`google-api-python-client`
` google-auth-httplib2`
 `google-auth-oauthlib`

**Comment by egede** [comment]
> These libraries will have to be added to https://github.com/ganga-devs/ganga/blob/develop/setup.py:
> `google-api-python-client`
> ` google-auth-httplib2`
> `google-auth-oauthlib`

OK, just go ahead and add them in as part of this PR.

**Review by egede — DISMISSED** [review]

**Comment by DumbMachine** [comment]
I believe that the work here is done, as far as the feature is concerned. Let me know if there is a need to replace the [Old Test](https://github.com/ganga-devs/ganga/blob/80bc4d9a6b5cb45bd72647236a782808bb4364e2/ganga/GangaCore/old_test/GPI/TestGoogleFile.py)

File: doc/UserGuide/InputAndOutputData.rst [hunk 1]
```diff
@@ -142,5 +142,30 @@ MassStorageFile
...
+.. code-block:: python
+
+    j = GangaFile("output.txt")
+    j.localDir = "~/temp"
+    j.put()
+
+This will upload the local file "~/temp/output.txt" to the user's Google Drive. The File object also supports for glob patterns, which can be supplied as `j.namePattern = '*.ROOT'`.
+Upon first usage, the user will be asked to authenticate and allow access to create new files and edit these files only.
+
+Files created by ganga can only be deleted ( or restored after deletion ).
```
Comment (egede) [inline doc/UserGuide/InputAndOutputData.rst]: I think you mean "Only files created by Ganga can be deleted (or restored after deletion)."

File: doc/UserGuide/InputAndOutputData.rst [hunk 2]
```diff
@@ -142,5 +142,30 @@ MassStorageFile
...
+To use a GangaFile, do something similar to:
+
+.. code-block:: python
+
+    j = GangaFile("output.txt")
+    j.localDir = "~/temp"
+    j.put()
+
+This will upload the local file "~/temp/output.txt" to the user's Google Drive. The File object also supports for glob patterns, which can be supplied as `j.namePattern = '*.ROOT'`.
+Upon first usage, the user will be asked to authenticate and allow access to create new files and edit these files only.
```
Comment (egede) [inline doc/UserGuide/InputAndOutputData.rst:166]: Can you add in the documentation how the URL for the file is obtained after upload.

**Review by egede — DISMISSED** [review]
A question here that I do not know the answer to. If I have a file in my GoogleDrive already, can I then do something like
```
gf = GoogleFile(....)
gf.get()
```
to get a local copy of it? If yes, please add an example of that in the docstring as well as the documentation.

**Comment by DumbMachine** [comment]
> A question here that I do not know the answer to. If I have a file in my GoogleDrive already, can I then do something like
> 
> ```
> gf = GoogleFile(....)
> gf.get()
> ```
> 
> to get a local copy of it? If yes, please add an example of that in the docstring as well as the documentation.

Currently, the `get` method is not implemented. But it can be done with ease if we have the user's authentication. I'll implement this and then add it to the documentation.

**Comment by egede** [comment]
When testing, I get the error below:
```
[23:41:15]                                                                                                                                                                                                         
Ganga In [5]: g.put()                                                                                                                                                                                              
ERROR    !!Unknown/Unexpected ERROR!!
ERROR    If you're able to reproduce this please report this to the Ganga developers!
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
~/programming/ganga/src/ganga/ganga/GangaCore/GPI/__init__.py in <module>
----> 1 g.put()

~/programming/ganga/src/ganga/ganga/GangaCore/GPIDev/Base/Proxy.py in proxy_wrapped(*args, **kwargs)
    718         s_kwargs = dict((name, stripProxy(a)) for name, a in kwargs.items())
    719         try:
--> 720             r = f(*s_args, **s_kwargs)
    721         except KeyboardInterrupt:
    722             logger.error("Command was interrupted by a Ctrl+C event!")

~/programming/ganga/src/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py in put(self)
    257         from apiclient.http import MediaFileUpload
    258 
--> 259         service = self._setup_service()
    260 
    261         # Sets the target directory

~/programming/ganga/src/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py in _setup_service(self)
    479         http = httplib2.Http()
    480         if self.__initialized == False:
--> 481             self.__initializeCred()
    482         with open(self.cred_path, "rb") as nput:
    483             credentials = pickle.load(nput)

~/programming/ganga/src/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py in __initializeCred(self)
    135         self.__initialized = True
    136 
--> 137         self._check_Ganga_folder()
    138 
    139     def _attribute_filter__set__(self, n, v):

~/programming/ganga/src/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py in _check_Ganga_folder(self)
    445         """
    446         from apiclient import errors
--> 447         service = self._setup_service()
    448 
    449         page_token = None

~/programming/ganga/src/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py in _setup_service(self)
    481             self.__initializeCred()
    482         with open(self.cred_path, "rb") as nput:
--> 483             credentials = pickle.load(nput)
    484         http = credentials.authorize(http)
    485         service = build('drive', 'v2', http=http)

ModuleNotFoundError: No module named 'oauth2client'
```

**Review by egede — CHANGES_REQUESTED** [review]
When I try to use the `get` command with a a file that was not uploaded by Ganga itself, I do not get an error. Just nothing happens.
```
[00:04:18]                                                                                                                                                                                                         
Ganga In [9]: g = GoogleFile()                                                                                                                                                                                     

[00:04:35]                                                                                                                                                                                                         
Ganga In [10]: g.namePattern='FileThatWasUploadedByOtherMethod'                                                                                                                                                                 

[00:04:47]                                                                                                                                                                                                         
Ganga In [11]: g.get()                                                                                                                                                                                             

[00:04:54]                                                                                                                                                                                                         
Ganga In [12]:                                                                                                                                                                                                     
```
A `GangaFileError` https://github.com/ganga-devs/ganga/blob/80e7769e6eb38d0ca98ccb8998b6e157860f8da9/ganga/GangaCore/Core/exceptions.py#L18 exception should be raised with a suitable error message explaining why the file cannot be downloaded.

**Comment by DumbMachine** [comment]
> When I try to use the `get` command with a a file that was not uploaded by Ganga itself, I do not get an error. Just nothing happens.
> 
> ```
> [00:04:18]                                                                                                                                                                                                         
> Ganga In [9]: g = GoogleFile()                                                                                                                                                                                     
> 
> [00:04:35]                                                                                                                                                                                                         
> Ganga In [10]: g.namePattern='FileThatWasUploadedByOtherMethod'                                                                                                                                                                 
> 
> [00:04:47]                                                                                                                                                                                                         
> Ganga In [11]: g.get()                                                                                                                                                                                             
> 
> [00:04:54]                                                                                                                                                                                                         
> Ganga In [12]:                                                                                                                                                                                                     
> ```
> 
> A `GangaFileError`
> 
> https://github.com/ganga-devs/ganga/blob/80e7769e6eb38d0ca98ccb8998b6e157860f8da9/ganga/GangaCore/Core/exceptions.py#L18
> 
> exception should be raised with a suitable error message explaining why the file cannot be downloaded.


What do you want to happen when the users uses `namePattern`  and among the whole bunch one file is missing. Should we just inform the user or raise `error`? Raising an `error` in this case would mean that the whole process would stop and exist with the error, this may lead to skipping further files.

We could make it that:
1. In case of single files, we raise an `error`
2. In case of multiple files, we can `inform` user about skipping that file since it was missing

**Comment by egede** [comment]
> What do you want to happen when the users uses namePattern and among the whole bunch one file is missing. Should we just inform the user or raise error? Raising an error in this case would mean that the whole process would stop and exist with the error, this may lead to skipping further files.

I think the default behaviour should be to raise the exception anyway. You can introduce a `keepgoing` argument to the `get` function.

```
def get(keepgoing=False)
```
where the behaviour is that with `keepgoing=False` the process is aborted and an exception raised on first failing download. With `keepgoing=True` no exceptions are raised and a WARNING is sent to the logger.

PS: Is some similar check and potential raise of an exception required for the put function as well?

**Comment by DumbMachine** [comment]
Agreed, I will add the `keepgoing` argument.
> PS: Is some similar check and potential raise of an exception required for the put function as well?

This is not required as the file will only be uploaded if it exists. This will happen when we try to upload a non-existent file:
```bash
[19:53:50]                                                                                                                                                     
Ganga In [6]: g = GoogleFile() 
         ...: g.namePattern = 'cntials.json' 
         ...: g.localDir = '~/temp'                                                                                                                            

[20:08:00]                                                                                                                                                     
Ganga In [7]: g.put()                                                                                                                                          
ERROR    !!Unknown/Unexpected ERROR!!
ERROR    If you're able to reproduce this please report this to the Ganga developers!
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
~/code/GSOC_SELECTION/ganga/ganga/GangaCore/GPI/__init__.py in <module>
----> 1 g.put()

~/code/GSOC_SELECTION/ganga/ganga/GangaCore/GPIDev/Base/Proxy.py in proxy_wrapped(*args, **kwargs)
    718         s_kwargs = dict((name, stripProxy(a)) for name, a in kwargs.items())
    719         try:
--> 720             r = f(*s_args, **s_kwargs)
    721         except KeyboardInterrupt:
    722             logger.error("Command was interrupted by a Ctrl+C event!")

~/code/GSOC_SELECTION/ganga/ganga/GangaCore/GPIDev/Lib/File/GoogleFile.py in put(self)
    343 
    344             file_metadata = {
--> 345                 'name': self.namePattern,
    346                 'description': 'A test document',
    347                 'mimeType': 'text/plain',

~/anaconda3/lib/python3.7/site-packages/googleapiclient/_helpers.py in positional_wrapper(*args, **kwargs)
    132                 elif positional_parameters_enforcement == POSITIONAL_WARNING:
    133                     logger.warning(message)
--> 134             return wrapped(*args, **kwargs)
    135 
    136         return positional_wrapper

~/anaconda3/lib/python3.7/site-packages/googleapiclient/http.py in __init__(self, filename, mimetype, chunksize, resumable)
    564     """
    565         self._filename = filename
--> 566         fd = open(self._filename, "rb")
    567         if mimetype is None:
    568             # No mimetype provided, make a guess.

FileNotFoundError: [Errno 2] No such file or directory: '/home/dumbmachine/temp/cntials.json'

```

**Comment by egede** [comment]
I would prefer that the exception above gets caught and then re-raised as a `GangaFileError` (with the original message). This is not a bug in Ganga. When raising a `GangaFileError` the user will get a clear error message but not a traceback.

**Comment by DumbMachine** [comment]
Upon some testing, I realized that there is no need for `keepgoing`. Using `namePatterns` the `get()` will now download all the files inside of `Ganga` folder that satisfy the pattern

**Review by egede — APPROVED** [review]
This is all good now. Nice work overall.

**Pull request merged**